File: src/android/app/src/main/java/com/wata/client/DmRoomService.kt
package com.wata.client

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

// ============================================================================
// DMRoomService: Encapsulates all DM room management logic
// ============================================================================

/**
 * DMRoomService: Encapsulates all DM room management logic
 *
 * This service handles:
 * - DM room lookup (userId -> roomId mapping)
 * - Deterministic room selection when multiple rooms exist
 * - m.direct account data management
 * - DM room creation (rare, only when no room exists)
 */
class DmRoomService(
    private val api: MatrixApi,
    private val syncEngine: SyncEngine,
    private val userId: String,
    logger: Logger? = null
) {
    private val logger: Logger = logger ?: createNoopLogger()

    /**
     * Primary room ID for each contact (deterministically selected).
     * This is THE room to use when sending messages.
     */
    private val primaryRoomByContact: MutableMap<String, String> = mutableMapOf()

    /**
     * All known room IDs for each contact (may be multiple due to race conditions).
     * Used for message consolidation and room detection.
     */
    private val allRoomsByContact: MutableMap<String, MutableSet<String>> = mutableMapOf()

    /**
     * Reverse lookup: room ID -> contact user ID.
     * Used for quick isDMRoom() checks and getContactForRoom().
     */
    private val contactByRoom: MutableMap<String, String> = mutableMapOf()

    // ==========================================================================
    // Lookup Methods (synchronous, cached)
    // ==========================================================================

    /**
     * Get the primary DM room ID for a contact (if known).
     * Returns null if no DM room exists in our cache.
     * Does NOT create a room or make network calls.
     */
    fun getDMRoomId(contactUserId: String): String? {
        return primaryRoomByContact[contactUserId]
    }

    /**
     * Get all known DM room IDs for a contact.
     * Returns empty list if none known.
     * Useful for message consolidation across duplicate rooms.
     */
    fun getAllDMRoomIds(contactUserId: String): List<String> {
        val roomIds = allRoomsByContact[contactUserId]
        return roomIds?.toList() ?: emptyList()
    }

    /**
     * Check if a room ID is a known DM room.
     */
    fun isDMRoom(roomId: String): Boolean {
        return contactByRoom.containsKey(roomId)
    }

    /**
     * Get the contact user ID for a DM room (reverse lookup).
     * Returns null if the room isn't a known DM room.
     */
    fun getContactUserId(roomId: String): String? {
        return contactByRoom[roomId]
    }

    /**
     * Get Contact object for a DM room.
     * Returns null if not a DM room or contact not found.
     */
    fun getContactForRoom(roomId: String): Contact? {
        // First try the cached mapping
        val contactUserId = contactByRoom[roomId]
        if (contactUserId != null) {
            // Cache hit - buildContactFromRoom always returns a Contact now
            return buildContactFromRoom(roomId, contactUserId)
        }

        // Fallback: Infer from room membership (recipient-side issue)
        // A DM room has exactly 2 members: current user and contact
        val room = syncEngine.getRoom(roomId)
        if (room == null) return null

        val joinedMembers = room.members.values.filter { it.membership == "join" }

        if (joinedMembers.size != 2) return null

        val otherMember = joinedMembers.find { it.userId != userId }
        if (otherMember == null) return null

        // Accept any 2-person room as a DM (relaxed check for recipient-side)
        // The is_direct flag may not be set on rooms created by others
        logger.log("[DMRoomService] getContactForRoom: inferred DM for ${otherMember.userId} from room membership (cache miss)")
        return Contact(
            user = User(
                id = otherMember.userId,
                displayName = otherMember.displayName,
                avatarUrl = otherMember.avatarUrl
            )
        )
    }

    // ==========================================================================
    // Creation Methods (async, makes network calls)
    // ==========================================================================

    /**
     * Ensure a DM room exists with the contact.
     * - First tries to find existing room via sync state
     * - Creates new room only if none found
     * - Updates m.direct and internal caches
     *
     * This is the ONLY method that should create DM rooms.
     */
    suspend fun ensureDMRoom(contactUserId: String): String {
        logger.log("[DMRoomService] ensureDMRoom for $contactUserId")

        // Step 1: Check cache first (fast path)
        val cached = primaryRoomByContact[contactUserId]
        if (cached != null) {
            // Verify the room still exists and we're joined
            val room = syncEngine.getRoom(cached)
            if (room != null && room.members[userId]?.membership == "join") {
                logger.log("[DMRoomService] Using cached room $cached")
                return cached
            }
            // Room no longer valid, remove from cache
            removeRoomFromCache(cached)
        }

        // Step 2: Scan sync state for existing DM rooms
        val existingRoom = findExistingDMRoom(contactUserId)
        if (existingRoom != null) {
            logger.log("[DMRoomService] Found existing room $existingRoom")
            updateMDirectForRoom(contactUserId, existingRoom)
            return existingRoom
        }

        // Step 3: No existing room found, create new one
        logger.log("[DMRoomService] Creating new DM room with $contactUserId")
        val roomId = createDMRoom(contactUserId)

        return roomId
    }

    // ==========================================================================
    // Cache Management
    // ==========================================================================

    /**
     * Handle m.direct account data update.
     * Called by WataClient when account data changes.
     */
    fun handleMDirectUpdate(content: Map<String, List<String>>) {
        // m.direct format: { "@user:server": ["!roomId1", "!roomId2", ...] }
        for ((contactUserId, roomIds) in content) {
            if (roomIds.isEmpty()) continue

            // Update allRoomsByContact
            allRoomsByContact[contactUserId] = roomIds.toMutableSet()

            // Update reverse mapping
            for (roomId in roomIds) {
                contactByRoom[roomId] = contactUserId
            }

            // Determine primary room (oldest by creation timestamp)
            val primaryRoomId = selectPrimaryRoom(roomIds)
            if (primaryRoomId != null) {
                primaryRoomByContact[contactUserId] = primaryRoomId
            }
        }
    }

    /**
     * Refresh cache from current sync state.
     * Called after sync batches to discover new DM rooms.
     */
    fun refreshFromSync() {
        val rooms = syncEngine.getRooms()
        logger.log("[DMRoomService] refreshFromSync: scanning ${rooms.size} rooms")

        for (room in rooms) {
            // Skip rooms we're not joined to
            val myMembership = room.members[userId]?.membership
            if (myMembership != "join") {
                logger.log("[DMRoomService] Room ${room.roomId.takeLast(12)}: skipping (my membership=$myMembership)")
                continue
            }

            // Check if this is a 2-person DM room
            val joinedMembers = room.members.values.filter { it.membership == "join" }
            if (joinedMembers.size != 2) {
                logger.log("[DMRoomService] Room ${room.roomId.takeLast(12)}: skipping (${joinedMembers.size} joined members)")
                continue
            }

            val otherMember = joinedMembers.find { it.userId != userId }
            if (otherMember == null) {
                logger.log("[DMRoomService] Room ${room.roomId.takeLast(12)}: skipping (no other member)")
                continue
            }

            // Verify is_direct flag
            if (!hasIsDirectFlag(room)) {
                logger.log("[DMRoomService] Room ${room.roomId.takeLast(12)}: skipping (no is_direct flag)")
                continue
            }

            // Add to cache
            val contactUserId = otherMember.userId
            logger.log("[DMRoomService] Room ${room.roomId.takeLast(12)}: detected as DM with $contactUserId")
            addRoomToCache(contactUserId, room.roomId)
        }
    }

    /**
     * Clear all cached state (on logout).
     */
    fun clear() {
        primaryRoomByContact.clear()
        allRoomsByContact.clear()
        contactByRoom.clear()
    }

    // ==========================================================================
    // Private Helpers
    // ==========================================================================

    /**
     * Find existing DM room with contact by scanning sync state.
     * Returns the primary room ID if found, null otherwise.
     */
    private fun findExistingDMRoom(contactUserId: String): String? {
        val candidateRooms = mutableListOf<CandidateRoom>()
        val rooms = syncEngine.getRooms()

        for (room in rooms) {
            // Skip if not joined
            if (room.members[userId]?.membership != "join") continue

            // Check if this is a 2-person room with the target user
            val joinedMembers = room.members.values.filter { it.membership == "join" }
            if (joinedMembers.size != 2) continue

            val hasTargetUser = joinedMembers.any { it.userId == contactUserId }
            if (!hasTargetUser) continue

            // Check is_direct flag and get creation timestamp
            val (isDirectRoom, creationTs) = getRoomDMInfo(room)

            // Only include valid DM rooms with timestamps
            if (isDirectRoom && creationTs != null && creationTs > 0) {
                val messageCount = room.timeline.count { event ->
                    event.type == "m.room.message" &&
                    event.content["msgtype"]?.jsonPrimitive?.content == "m.audio"
                }
                candidateRooms.add(CandidateRoom(room.roomId, creationTs, messageCount))
                logger.log(
                    "[DMRoomService] Found candidate room ${room.roomId} " +
                    "(created: $creationTs, $messageCount msgs)"
                )
            }
        }

        if (candidateRooms.isEmpty()) return null

        // Log warning if multiple rooms found
        if (candidateRooms.size > 1) {
            val roomList = candidateRooms.joinToString(", ") { r ->
                "${r.roomId.takeLast(12)} (${r.creationTs}, ${r.messageCount} msgs)"
            }
            logger.warn(
                "[DMRoomService] Multiple DM rooms with $contactUserId: $roomList. Selecting oldest."
            )
        }

        // Sort by creation timestamp (oldest first), room ID as tiebreaker
        candidateRooms.sortWith(compareBy({ it.creationTs }, { it.roomId }))

        val primaryRoom = candidateRooms[0]
        logger.log(
            "[DMRoomService] Selected primary room ${primaryRoom.roomId} " +
            "(${candidateRooms.size} candidates)"
        )

        // Update cache with all candidate rooms
        for (room in candidateRooms) {
            addRoomToCache(contactUserId, room.roomId)
        }
        primaryRoomByContact[contactUserId] = primaryRoom.roomId

        return primaryRoom.roomId
    }

    /**
     * Create a new DM room with the contact.
     */
    private suspend fun createDMRoom(contactUserId: String): String {
        val response = api.createRoom(
            CreateRoomRequest(
                is_direct = true,
                invite = listOf(contactUserId),
                preset = "trusted_private_chat",
                visibility = "private"
            )
        )

        val roomId = response.room_id

        // Update m.direct account data
        updateMDirectForRoom(contactUserId, roomId)

        // Add to cache
        addRoomToCache(contactUserId, roomId)
        primaryRoomByContact[contactUserId] = roomId

        return roomId
    }

    /**
     * Update m.direct account data with a DM room.
     */
    private suspend fun updateMDirectForRoom(contactUserId: String, roomId: String) {
        try {
            // Get current m.direct data (map of userId -> list of roomIds)
            val directData: MutableMap<String, MutableList<String>> = mutableMapOf()
            try {
                val accountData = api.getAccountData(userId, "m.direct")
                // Parse: { "@user:server": ["!roomId1", "!roomId2"], ... }
                for ((key, value) in accountData.content) {
                    val roomIds = value.jsonArray.map { it.jsonPrimitive.content }
                    directData[key] = roomIds.toMutableList()
                }
            } catch (e: Exception) {
                // No existing data - start fresh
            }

            // Add room to contact's list if not already there
            val roomList = directData.getOrPut(contactUserId) { mutableListOf() }
            if (!roomList.contains(roomId)) {
                roomList.add(roomId)

                // Serialize back to JsonObject and persist
                val newContent = buildJsonObject {
                    for ((key, rooms) in directData) {
                        put(key, JsonArray(rooms.map { kotlinx.serialization.json.JsonPrimitive(it) }))
                    }
                }
                api.setAccountData(userId, "m.direct", newContent)
                logger.log("[DMRoomService] Updated m.direct with room $roomId for $contactUserId")
            }
        } catch (e: Exception) {
            logger.error("[DMRoomService] Failed to update m.direct: $e")
        }
    }

    /**
     * Select primary room from a list of room IDs (oldest wins).
     */
    private fun selectPrimaryRoom(roomIds: List<String>): String? {
        val roomsWithTs = mutableListOf<RoomWithTimestamp>()

        for (roomId in roomIds) {
            val room = syncEngine.getRoom(roomId)
            if (room != null) {
                // Get creation timestamp
                for (event in room.timeline) {
                    if (event.type == "m.room.create" && event.origin_server_ts != null) {
                        roomsWithTs.add(RoomWithTimestamp(roomId, event.origin_server_ts))
                        break
                    }
                }
            }
        }

        if (roomsWithTs.isEmpty()) {
            // Fall back to first room if no timestamps available
            return roomIds.firstOrNull()
        }

        // Sort by creation timestamp (oldest first)
        roomsWithTs.sortBy { it.creationTs }
        return roomsWithTs[0].roomId
    }

    /**
     * Add a room to the cache for a contact.
     */
    private fun addRoomToCache(contactUserId: String, roomId: String) {
        // Update allRoomsByContact
        if (!allRoomsByContact.containsKey(contactUserId)) {
            allRoomsByContact[contactUserId] = mutableSetOf()
        }
        allRoomsByContact[contactUserId]!!.add(roomId)

        // Update reverse mapping
        contactByRoom[roomId] = contactUserId
    }

    /**
     * Remove a room from all caches.
     */
    private fun removeRoomFromCache(roomId: String) {
        val contactUserId = contactByRoom[roomId]
        if (contactUserId == null) return

        // Remove from allRoomsByContact
        allRoomsByContact[contactUserId]?.remove(roomId)
        if (allRoomsByContact[contactUserId]?.isEmpty() == true) {
            allRoomsByContact.remove(contactUserId)
        }

        // Remove from primaryRoomByContact if it's the primary
        if (primaryRoomByContact[contactUserId] == roomId) {
            primaryRoomByContact.remove(contactUserId)
        }

        // Remove reverse mapping
        contactByRoom.remove(roomId)
    }

    /**
     * Check if a room has the is_direct flag set.
     *
     * Note: is_direct is a boolean in JSON ({"is_direct": true}), not a string.
     * We need to check booleanOrNull, not content (which would give "true" string).
     */
    private fun hasIsDirectFlag(room: SyncRoomState): Boolean {
        for (event in room.timeline) {
            val isDirect = event.content["is_direct"]?.jsonPrimitive?.booleanOrNull
            if (event.type == "m.room.create" && isDirect == true) {
                return true
            }
            if (event.type == "m.room.member" &&
                event.state_key == userId &&
                isDirect == true) {
                return true
            }
        }
        return false
    }

    /**
     * Get DM-related info from a room (is_direct flag and creation timestamp).
     *
     * Note: is_direct is a boolean in JSON ({"is_direct": true}), not a string.
     */
    private fun getRoomDMInfo(room: SyncRoomState): RoomDMInfo {
        var isDirectRoom = false
        var creationTs: Long? = null

        for (event in room.timeline) {
            val isDirect = event.content["is_direct"]?.jsonPrimitive?.booleanOrNull
            if (event.type == "m.room.create") {
                creationTs = event.origin_server_ts
                if (isDirect == true) {
                    isDirectRoom = true
                }
            }
            if (event.type == "m.room.member" &&
                event.state_key == userId &&
                isDirect == true) {
                isDirectRoom = true
            }
        }

        return RoomDMInfo(isDirectRoom, creationTs)
    }

    /**
     * Build a Contact object from room membership info.
     * Works for joined members as well as invited members (DM room with pending invite).
     */
    private fun buildContactFromRoom(roomId: String, contactUserId: String): Contact? {
        val room = syncEngine.getRoom(roomId)
        if (room == null) {
            return null
        }

        val member = room.members[contactUserId]
        if (member != null) {
            return Contact(
                user = User(
                    id = contactUserId,
                    displayName = member.displayName,
                    avatarUrl = member.avatarUrl
                )
            )
        }

        // Member not in room members - use contactUserId as fallback
        // This handles cases where the invite hasn't propagated yet
        val displayName = contactUserId.substringBefore(":").removePrefix("@")
        return Contact(
            user = User(
                id = contactUserId,
                displayName = displayName,
                avatarUrl = null
            )
        )
    }
}

// ============================================================================
// Internal Data Classes
// ============================================================================

private data class CandidateRoom(
    val roomId: String,
    val creationTs: Long,
    val messageCount: Int
)

private data class RoomWithTimestamp(
    val roomId: String,
    val creationTs: Long
)

private data class RoomDMInfo(
    val isDirectRoom: Boolean,
    val creationTs: Long?
)

// ============================================================================
// No-op Logger
// ============================================================================

private fun createNoopLogger(): Logger {
    return object : Logger {
        override fun log(message: String) {}
        override fun warn(message: String) {}
        override fun error(message: String) {}
    }
}

File: src/shared/lib/wata-client/dm-room-service.ts
/**
 * DMRoomService: Encapsulates all DM room management logic
 *
 * This service handles:
 * - DM room lookup (userId → roomId mapping)
 * - Deterministic room selection when multiple rooms exist
 * - m.direct account data management
 * - DM room creation (rare, only when no room exists)
 *
 * See PLAN.md for design rationale.
 */

import type { MatrixApi } from './matrix-api';
import type { SyncEngine, RoomState } from './sync-engine';
import type { Logger, Contact } from './types';

// ============================================================================
// No-op logger (default when no logger provided)
// ============================================================================

const noopLogger: Logger = {
  log: () => {},
  warn: () => {},
  error: () => {},
};

// ============================================================================
// DMRoomService Implementation
// ============================================================================

export class DMRoomService {
  private api: MatrixApi;
  private syncEngine: SyncEngine;
  private userId: string;
  private logger: Logger;

  /**
   * Primary room ID for each contact (deterministically selected).
   * This is THE room to use when sending messages.
   */
  private primaryRoomByContact: Map<string, string> = new Map();

  /**
   * All known room IDs for each contact (may be multiple due to race conditions).
   * Used for message consolidation and room detection.
   */
  private allRoomsByContact: Map<string, Set<string>> = new Map();

  /**
   * Reverse lookup: room ID → contact user ID.
   * Used for quick isDMRoom() checks and getContactForRoom().
   */
  private contactByRoom: Map<string, string> = new Map();

  constructor(
    api: MatrixApi,
    syncEngine: SyncEngine,
    userId: string,
    logger?: Logger
  ) {
    this.api = api;
    this.syncEngine = syncEngine;
    this.userId = userId;
    this.logger = logger ?? noopLogger;
  }

  // ==========================================================================
  // Lookup Methods (synchronous, cached)
  // ==========================================================================

  /**
   * Get the primary DM room ID for a contact (if known).
   * Returns null if no DM room exists in our cache.
   * Does NOT create a room or make network calls.
   */
  getDMRoomId(contactUserId: string): string | null {
    return this.primaryRoomByContact.get(contactUserId) ?? null;
  }

  /**
   * Get all known DM room IDs for a contact.
   * Returns empty array if none known.
   * Useful for message consolidation across duplicate rooms.
   */
  getAllDMRoomIds(contactUserId: string): string[] {
    const roomIds = this.allRoomsByContact.get(contactUserId);
    return roomIds ? Array.from(roomIds) : [];
  }

  /**
   * Check if a room ID is a known DM room.
   */
  isDMRoom(roomId: string): boolean {
    return this.contactByRoom.has(roomId);
  }

  /**
   * Get the contact user ID for a DM room (reverse lookup).
   * Returns null if the room isn't a known DM room.
   */
  getContactUserId(roomId: string): string | null {
    return this.contactByRoom.get(roomId) ?? null;
  }

  /**
   * Get Contact object for a DM room.
   * Returns null if not a DM room or contact not found.
   */
  getContactForRoom(roomId: string): Contact | null {
    // First try the cached mapping
    const contactUserId = this.contactByRoom.get(roomId);
    if (contactUserId) {
      // Cache hit - buildContactFromRoom always returns a Contact now
      return this.buildContactFromRoom(roomId, contactUserId);
    }

    // Fallback: Infer from room membership (recipient-side issue)
    // A DM room has exactly 2 members: current user and contact
    const room = this.syncEngine.getRoom(roomId);
    if (!room) return null;

    const joinedMembers = Array.from(room.members.values()).filter(
      (m) => m.membership === 'join'
    );

    if (joinedMembers.length !== 2) return null;

    const otherMember = joinedMembers.find((m) => m.userId !== this.userId);
    if (!otherMember) return null;

    // Verify this looks like a DM by checking is_direct flag
    if (!this.hasIsDirectFlag(room)) return null;

    return {
      user: {
        id: otherMember.userId,
        displayName: otherMember.displayName,
        avatarUrl: otherMember.avatarUrl,
      },
    };
  }

  // ==========================================================================
  // Creation Methods (async, makes network calls)
  // ==========================================================================

  /**
   * Ensure a DM room exists with the contact.
   * - First tries to find existing room via sync state
   * - Creates new room only if none found
   * - Updates m.direct and internal caches
   *
   * This is the ONLY method that should create DM rooms.
   */
  async ensureDMRoom(contactUserId: string): Promise<string> {
    this.logger.log(`[DMRoomService] ensureDMRoom for ${contactUserId}`);

    // Step 1: Check cache first (fast path)
    const cached = this.primaryRoomByContact.get(contactUserId);
    if (cached) {
      // Verify the room still exists and we're joined
      const room = this.syncEngine.getRoom(cached);
      if (room && room.members.get(this.userId)?.membership === 'join') {
        this.logger.log(`[DMRoomService] Using cached room ${cached}`);
        return cached;
      }
      // Room no longer valid, remove from cache
      this.removeRoomFromCache(cached);
    }

    // Step 2: Scan sync state for existing DM rooms
    const existingRoom = this.findExistingDMRoom(contactUserId);
    if (existingRoom) {
      this.logger.log(`[DMRoomService] Found existing room ${existingRoom}`);
      await this.updateMDirectForRoom(contactUserId, existingRoom);
      return existingRoom;
    }

    // Step 3: No existing room found, create new one
    this.logger.log(`[DMRoomService] Creating new DM room with ${contactUserId}`);
    const roomId = await this.createDMRoom(contactUserId);

    return roomId;
  }

  // ==========================================================================
  // Cache Management
  // ==========================================================================

  /**
   * Handle m.direct account data update.
   * Called by WataClient when account data changes.
   */
  handleMDirectUpdate(content: Record<string, string[]>): void {
    // m.direct format: { "@user:server": ["!roomId1", "!roomId2", ...] }
    for (const [contactUserId, roomIds] of Object.entries(content)) {
      if (!Array.isArray(roomIds) || roomIds.length === 0) continue;

      // Update all rooms mapping
      this.allRoomsByContact.set(contactUserId, new Set(roomIds));

      // Update reverse mapping
      for (const roomId of roomIds) {
        this.contactByRoom.set(roomId, contactUserId);
      }

      // Determine primary room (oldest by creation timestamp)
      const primaryRoomId = this.selectPrimaryRoom(roomIds);
      if (primaryRoomId) {
        this.primaryRoomByContact.set(contactUserId, primaryRoomId);
      }
    }
  }

  /**
   * Refresh cache from current sync state.
   * Called after sync batches to discover new DM rooms.
   */
  refreshFromSync(): void {
    const rooms = this.syncEngine.getRooms();

    for (const room of rooms) {
      // Skip rooms we're not joined to
      if (room.members.get(this.userId)?.membership !== 'join') continue;

      // Check if this is a 2-person DM room
      const joinedMembers = Array.from(room.members.values()).filter(
        (m) => m.membership === 'join'
      );
      if (joinedMembers.length !== 2) continue;

      const otherMember = joinedMembers.find((m) => m.userId !== this.userId);
      if (!otherMember) continue;

      // Verify is_direct flag
      if (!this.hasIsDirectFlag(room)) continue;

      // Add to cache
      const contactUserId = otherMember.userId;
      this.addRoomToCache(contactUserId, room.roomId);
    }
  }

  /**
   * Clear all cached state (on logout).
   */
  clear(): void {
    this.primaryRoomByContact.clear();
    this.allRoomsByContact.clear();
    this.contactByRoom.clear();
  }

  // ==========================================================================
  // Private Helpers
  // ==========================================================================

  /**
   * Find existing DM room with contact by scanning sync state.
   * Returns the primary room ID if found, null otherwise.
   */
  private findExistingDMRoom(contactUserId: string): string | null {
    const candidateRooms: { roomId: string; creationTs: number; messageCount: number }[] = [];
    const rooms = this.syncEngine.getRooms();

    for (const room of rooms) {
      // Skip if not joined
      if (room.members.get(this.userId)?.membership !== 'join') continue;

      // Check if this is a 2-person room with the target user
      const joinedMembers = Array.from(room.members.values()).filter(
        (m) => m.membership === 'join'
      );
      if (joinedMembers.length !== 2) continue;

      const hasTargetUser = joinedMembers.some((m) => m.userId === contactUserId);
      if (!hasTargetUser) continue;

      // Check is_direct flag and get creation timestamp
      const { isDirectRoom, creationTs } = this.getRoomDMInfo(room);

      // Only include valid DM rooms with timestamps
      if (isDirectRoom && creationTs !== null && creationTs > 0) {
        const messageCount = room.timeline.filter(
          (e) => e.type === 'm.room.message' && e.content?.msgtype === 'm.audio'
        ).length;
        candidateRooms.push({ roomId: room.roomId, creationTs, messageCount });
        this.logger.log(
          `[DMRoomService] Found candidate room ${room.roomId} (created: ${new Date(creationTs).toISOString()}, ${messageCount} msgs)`
        );
      }
    }

    if (candidateRooms.length === 0) return null;

    // Log warning if multiple rooms found
    if (candidateRooms.length > 1) {
      const roomList = candidateRooms
        .map((r) => `${r.roomId.slice(-12)} (${new Date(r.creationTs).toISOString().slice(0, 10)}, ${r.messageCount} msgs)`)
        .join(', ');
      this.logger.warn(
        `[DMRoomService] Multiple DM rooms with ${contactUserId}: ${roomList}. Selecting oldest.`
      );
    }

    // Sort by creation timestamp (oldest first), room ID as tiebreaker
    candidateRooms.sort((a, b) => {
      if (a.creationTs !== b.creationTs) return a.creationTs - b.creationTs;
      return a.roomId.localeCompare(b.roomId);
    });

    const primaryRoom = candidateRooms[0];
    this.logger.log(
      `[DMRoomService] Selected primary room ${primaryRoom.roomId} (${candidateRooms.length} candidates)`
    );

    // Update cache with all candidate rooms
    for (const room of candidateRooms) {
      this.addRoomToCache(contactUserId, room.roomId);
    }
    this.primaryRoomByContact.set(contactUserId, primaryRoom.roomId);

    return primaryRoom.roomId;
  }

  /**
   * Create a new DM room with the contact.
   */
  private async createDMRoom(contactUserId: string): Promise<string> {
    const response = await this.api.createRoom({
      is_direct: true,
      invite: [contactUserId],
      preset: 'trusted_private_chat',
      visibility: 'private',
    });

    const roomId = response.room_id;

    // Update m.direct account data
    await this.updateMDirectForRoom(contactUserId, roomId);

    // Add to cache
    this.addRoomToCache(contactUserId, roomId);
    this.primaryRoomByContact.set(contactUserId, roomId);

    return roomId;
  }

  /**
   * Update m.direct account data with a DM room.
   */
  private async updateMDirectForRoom(
    contactUserId: string,
    roomId: string
  ): Promise<void> {
    try {
      // Get current m.direct data
      let directData: Record<string, string[]> = {};
      try {
        directData = await this.api.getAccountData(this.userId, 'm.direct');
      } catch {
        // No existing data
      }

      // Add room to contact's list if not already there
      if (!directData[contactUserId]) {
        directData[contactUserId] = [];
      }
      if (!directData[contactUserId].includes(roomId)) {
        directData[contactUserId].push(roomId);
        await this.api.setAccountData(this.userId, 'm.direct', directData);
      }
    } catch (error) {
      this.logger.error(`[DMRoomService] Failed to update m.direct: ${error}`);
    }
  }

  /**
   * Select primary room from a list of room IDs (oldest wins).
   */
  private selectPrimaryRoom(roomIds: string[]): string | null {
    const roomsWithTs: { roomId: string; creationTs: number }[] = [];

    for (const roomId of roomIds) {
      const room = this.syncEngine.getRoom(roomId);
      if (!room) continue;

      // Get creation timestamp
      for (const event of room.timeline) {
        if (event.type === 'm.room.create' && event.origin_server_ts) {
          roomsWithTs.push({ roomId, creationTs: event.origin_server_ts });
          break;
        }
      }
    }

    if (roomsWithTs.length === 0) {
      // Fall back to first room if no timestamps available
      return roomIds[0] ?? null;
    }

    // Sort by creation timestamp (oldest first)
    roomsWithTs.sort((a, b) => a.creationTs - b.creationTs);
    return roomsWithTs[0].roomId;
  }

  /**
   * Add a room to the cache for a contact.
   */
  private addRoomToCache(contactUserId: string, roomId: string): void {
    // Update allRoomsByContact
    if (!this.allRoomsByContact.has(contactUserId)) {
      this.allRoomsByContact.set(contactUserId, new Set());
    }
    this.allRoomsByContact.get(contactUserId)!.add(roomId);

    // Update reverse mapping
    this.contactByRoom.set(roomId, contactUserId);
  }

  /**
   * Remove a room from all caches.
   */
  private removeRoomFromCache(roomId: string): void {
    const contactUserId = this.contactByRoom.get(roomId);
    if (!contactUserId) return;

    // Remove from allRoomsByContact
    const rooms = this.allRoomsByContact.get(contactUserId);
    if (rooms) {
      rooms.delete(roomId);
      if (rooms.size === 0) {
        this.allRoomsByContact.delete(contactUserId);
      }
    }

    // Remove from primaryRoomByContact if it's the primary
    if (this.primaryRoomByContact.get(contactUserId) === roomId) {
      this.primaryRoomByContact.delete(contactUserId);
    }

    // Remove reverse mapping
    this.contactByRoom.delete(roomId);
  }

  /**
   * Check if a room has the is_direct flag set.
   */
  private hasIsDirectFlag(room: RoomState): boolean {
    // Check if our own member event has is_direct flag (set by server when room is created as DM)
    const myMember = room.members.get(this.userId);
    if (myMember?.isDirect === true) {
      return true;
    }

    // Fallback: A 2-member room (current user + one other) is a DM
    // This handles rooms where we joined (not created via is_direct invite)
    const joinedMembers = Array.from(room.members.values()).filter(
      (m) => m.membership === 'join'
    );
    if (joinedMembers.length === 2) {
      const otherMember = joinedMembers.find((m) => m.userId !== this.userId);
      if (otherMember) {
        return true;
      }
    }

    // Timeline scan for is_direct (for legacy deduplication)
    for (const event of room.timeline) {
      if (
        event.type === 'm.room.member' &&
        event.state_key === this.userId &&
        event.content?.is_direct === true
      ) {
        return true;
      }
    }
    return false;
  }

  /**
   * Get DM-related info from a room (is_direct flag and creation timestamp).
   */
  private getRoomDMInfo(room: RoomState): { isDirectRoom: boolean; creationTs: number | null } {
    let isDirectRoom = false;
    let creationTs: number | null = null;

    for (const event of room.timeline) {
      if (event.type === 'm.room.create') {
        creationTs = event.origin_server_ts ?? null;
        if (event.content?.is_direct === true) {
          isDirectRoom = true;
        }
      }
      if (
        event.type === 'm.room.member' &&
        event.state_key === this.userId &&
        event.content?.is_direct === true
      ) {
        isDirectRoom = true;
      }
    }

    return { isDirectRoom, creationTs };
  }

  /**
   * Build a Contact object from room membership info.
   * Works for joined members as well as invited members (DM room with pending invite).
   */
  private buildContactFromRoom(roomId: string, contactUserId: string): Contact | null {
    const room = this.syncEngine.getRoom(roomId);
    if (!room) {
      return null;
    }

    const member = room.members.get(contactUserId);
    if (member) {
      return {
        user: {
          id: contactUserId,
          displayName: member.displayName,
          avatarUrl: member.avatarUrl,
        },
      };
    }

    // Member not in room members - use contactUserId as fallback
    // This handles cases where the invite hasn't propagated yet
    return {
      user: {
        id: contactUserId,
        displayName: contactUserId.split(':')[0].substring(1), // @bob:server -> bob
        avatarUrl: null,
      },
    };
  }
}
